### ISSUE TITLE
some images in epubs are not saved

### ISSUE THREAD

**Issue by stefan5588** [danieldeng2/perlego-downloader#13]
Hi, thanks for this script, it works great! But in epubs, there are still some images which don't get saved - try to download this page for instance: https://ereader.perlego.com/1/book/2322378/4

**Comment by danieldeng2** [comment]
Should be fixed. Try again.

**Comment by stefan5588** [comment]
Sorry to say, but it's still not working with the example given above. On https://ereader.perlego.com/1/book/2322378/4 it should be an image right below the "First Movement" header  - the URL of the image is https://static.perlego.com/v4/image?f=31b8f6cc2a37fb2adfe057e1f5ee32577f953f74b5563&amp;q=db0b3b67110a8e2bac8af9dcc32f67db87a56eaa861cc&#39;). Yet it's not loading when running your updated script.

**Comment by stefan5588** [comment]
The problem I think is with the element selection on that specific page - your script assumes there is only one "data-chapterid" per page, which is probably true most of the time - yet the page above has TWO chapterid on the same page - one has the "First Movement" text, the other one has the image. So, the script should be modified to account for these cases with multiples "chapterid" per page.

**Comment by danieldeng2** [comment]
If you submit a pull request, I'll be happy to merge it

**Comment by stefan5588** [comment]
Well, I would submit a pull request if I would know how to change the code :) But all I know is the cause behind the issue.

**Comment by danieldeng2** [comment]
Unfortunately, I cannot test the book you referenced since it's not available in my country. 

The issue you mentioned should be able to be solved by making modifications to the loop, but it would be difficult for me to verify that it's working.

**Comment by stefan5588** [comment]
I will see if I can find another example. If you have time, here (https://we.tl/t-qnBoz99vSG) is a small archive containing the above page in HTML - you have both the original HTML ("OriginalLayout") as displayed on Perlego, and the HTML of the page as saved by your script ("ScriptLayout"). If you look at the HTML source, you'll see that in OriginalLayout there is both "data-chapterid="4"" (= text) and "data-chapterid="5" (= image), while the ScriptLayout retains only the first chapterid (""data-chapterid="4"").

**Comment by stefan5588** [comment]
Another example: https://ereader.perlego.com/1/book/2112289/1. This one has no less than 17 (!!) different "chapterid" on the same page, and the script only reads the first one ("data-chapterid="1""). This type of HTML page layout (miss-)mapping an EPUB, with multiple "page/chapter IDs" on the same page, is typical of some recent titles on Perlego, and typical of some publishing houses (Bloomsbury, Routledge). So, once again, we must find a way to capture multiple page/chapter IDs on the same HTML page and, as you said, modify the loop accordingly. Your script looks, for element selection, at "data-chapterid" - I wonder if it wouldn't be easier to look instead at "id="p1--0"", "id="p2--0"", "id="p3--0"", etc., since these values appear only once in HTML?

**Comment by stefan5588** [comment]
I think I've solved the issue by changing the element selection from "document.querySelector(`div[data-chapterid='${j}']`)" to "document.querySelector(`div[id='content']`)". But this works only for EPUBS, it breaks the processing of PDFs...

**Comment by danieldeng2** [comment]
I thought of that as well🙁, if it's a common issue we could separate pdf and epub into two different functions.

**Comment by stefan5588** [comment]
Yes, perhaps a constant "isEPUB" similar to "isPDF"? Anyway, there is another issue: while the "(div[id='content'])" element selection works well for EPUBs, some ill-formed HTML layouts have a "next chapter" button at the very end of the document. For instance, the above quoted page (with 17 chapterid on the same page), https://ereader.perlego.com/1/book/2112289/1, should end there, at page 1, but it has a "next chapter" button - so, if you run the script (with the modified element selection), you'll end up with 16 copies of the last page... So we need to find another way to end the loop.
